### PR TITLE
[Authn] Support AAD and new user & group schema

### PR DESCRIPTION
**Pull request by ydye**

### CODE REVIEW

File: src/webportal/src/app/home/index.jsx [hunk 1]
```diff
@@ -46,62 +58,68 @@ const Index = () => {
...
+      [],
   );
 
   const dismissLoginModal = useCallback(
-    () => {
-      setLoginModal(false);
-      setError(null);
-    },
-    [],
+      () => {
```
Comment (sunqinzheng) [inline src/webportal/src/app/home/index.jsx]: Please fix the indent

File: src/rest-server/package.json [hunk 2]
```diff
@@ -58,7 +59,11 @@
     "statuses": "~1.5.0",
     "unirest": "~0.5.1",
     "winston": "~2.4.0",
-    "xml2js": "~0.4.19"
+    "xml2js": "~0.4.19",
+    "passport": "*",
```
Comment (sunqinzheng) [inline src/rest-server/package.json]: Please remove unused dependencies.

File: src/rest-server/src/config/authn.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  authnConfig.OIDCConfig = yaml.safeLoad(fs.readFileSync('/auth-configuration/oidc.yaml', 'utf8'));
+}
+
+try {
+  authnConfig.groupConfig = yaml.safeLoad(fs.readFileSync('/group-configuration/group.yaml', 'utf8'));
+} catch (error) {
+  logger.error('Failed to load group config from configmap file.');
+  if (process.env.NODE_ENV === 'test') {
+    logger.error('Init groupConfig with default configuration. ');
+    authnConfig.groupConfig = {
```
Comment (sunqinzheng) [inline src/rest-server/src/config/authn.js]: Please move the test config initialization logic into test folder

File: src/rest-server/src/controllers/job.js [hunk 4]
```diff
@@ -120,6 +140,51 @@ const update = (req, res, next) => {
...
+        res(job);
+      }
+    });
+  });
+};
+
+/**
+ * Async API. Submit or update job
+ */
+const updateAsync = async (req, res, next) => {
```
Comment (sunqinzheng) [inline src/rest-server/src/controllers/job.js]: Please remove old job update API.

File: src/rest-server/src/controllers/v2/azureAD.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+const requestAuthCode = async (req, res, next) => {
+  const tenantId = authnConfig.OIDCConfig.tenantID;
+  const clientId = authnConfig.OIDCConfig.clientID;
+  const responseType = 'code';
+  const redirectUri = authnConfig.OIDCConfig.redirectUrl;
+  const responseMode = 'form_post';
+  const scope = 'openid offline_access https://graph.microsoft.com/user.read';
+  const state = 'openpai';
+  return res.redirect('https://login.microsoftonline.com/' + tenantId + '/oauth2/v2.0/authorize?'+ querystring.stringify({
```
Comment (sunqinzheng) [inline src/rest-server/src/controllers/v2/azureAD.js]: I think we'd better use OIDC's well known config to get the authroize and token url.
Some Azure region may not use `https://login.microsoftonline.com/` as the login host.
e.g. mooncake: https://docs.azure.cn/zh-cn/active-directory/develop/active-directory-v2-protocols

Comment (sunqinzheng) [reply src/rest-server/src/controllers/v2/azureAD.js]: Graph API endpoint is also different. https://docs.azure.cn/zh-cn/active-directory/develop/active-directory-graph-api-quickstart#how-to-construct-a-graph-api-url

Comment (sunqinzheng) [reply src/rest-server/src/controllers/v2/azureAD.js]: I think we should create a known issue if we only supports the default endpoint.

Comment (sunqinzheng) [reply src/rest-server/src/controllers/v2/azureAD.js]: National Clouds endpoint list https://docs.microsoft.com/en-us/azure/active-directory/develop/authentication-national-cloud

File: src/webportal/src/app/vc/vc.component.scss [hunk 6]
```diff
@@ -101,4 +101,4 @@ ul, li{
     color: #c0c0c0;
     cursor: default;
   }
-}
+}
```
Comment (sunqinzheng) [inline src/webportal/src/app/vc/vc.component.scss]: why the newline at end of file is removed?

File: src/webportal/src/app/user/fabric/batchRegister/index.jsx [hunk 7]
```diff
@@ -200,34 +204,17 @@ export default function BatchRegister() {
 
       let result = await createUserRequest(
         userInfo[columnUsername],
+        userInfo[columnEmail],
         userInfo[columnPassword],
-        toBool(userInfo[columnAdmin])).then(() => {
+        toBool(userInfo[columnAdmin]),
+        userInfo.vcs).then(() => {
```
Comment (sunqinzheng) [inline src/webportal/src/app/user/fabric/batchRegister/index.jsx]: indent

File: src/webportal/src/app/user/fabric/userView/utils.js [hunk 8]
```diff
@@ -15,10 +15,6 @@
 // DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-export const toBool = (val) => {
-  return (String(val)).toLowerCase() === 'true' || (String(val)).toLowerCase() === 'yes';
-};
-
 export const getVirtualCluster = (user) => {
-  return toBool(user.admin) ? 'All virtual clusters' : user.virtualCluster;
+  return user.admin ? 'All virtual clusters' : user.virtualCluster.slice().sort().join(', ');
```
Comment (sunqinzheng) [inline src/webportal/src/app/user/fabric/userView/utils.js]: user.virtualCluster used here, but home page uses user.groupList.
I think we should unify the properties.

File: src/webportal/src/app/job/job-view/fabric/JobList/TopBar.jsx [hunk 9]
```diff
@@ -1,22 +1,5 @@
-// Copyright (c) Microsoft Corporation
-// All rights reserved.
-//
-// MIT License
-//
-// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
-// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
```
Comment (sunqinzheng) [inline src/webportal/src/app/job/job-view/fabric/JobList/TopBar.jsx]: I think the license header should be keeped.

File: src/rest-server/src/routes/v2/user.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+const jobRouter = require('../job');
+const authnConfig = require('../../config/authn');
+
+const router = new express.Router();
+
+router.route('/get/:username/')
+  /** Get /api/v2/user/get/:username */
+  .get(token.check, userController.getUser);
+
+router.route('/get/')
```
Comment (sunqinzheng) [inline src/rest-server/src/routes/v2/user.js]: The `/get/` verb should not be included in rest api's url

File: src/rest-server/src/controllers/v2/token.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+const getAAD = async (req, res, next) => {
+  try {
+    const username = req.username;
+    const userInfo = await userModel.getUser(username);
+    let admin = false;
+    if (userInfo.grouplist.includes(authConfig.groupConfig.adminGroup.groupname)) {
+      admin = true;
+    }
+    const token = await jwtSignPromise(userInfo, admin);
+    return res.redirect('http://' + process.env.WEBPORTAL_URL + '/index.html?'+ querystring.stringify({
```
Comment (sunqinzheng) [inline src/rest-server/src/controllers/v2/token.js]: I think we should get the final redirect url from the initial token request

File: src/rest-server/src/routes/authn.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    /** POST /api/v1/authn/oidc/return - AAD AUTH RETURN */
+    .post(
+      azureADController.requestTokenWithCode,
+      azureADController.parseTokenData,
+      userController.createUserIfUserNotExist,
+      userController.updateUserGroupListFromExternal,
+      tokenV2Controller.getAAD
+    );
+} else {
+  router.route('/basic/login')
```
Comment (sunqinzheng) [inline src/rest-server/src/routes/authn.js]: both oidc and basic auth should use same endpoint with diffenrent behaviors

File: src/rest-server/package.json (hunk 2 above)
Comment (ydye) [reply src/rest-server/package.json]: Done

File: src/webportal/src/app/vc/vc.component.scss (hunk 6 above)
Comment (ydye) [reply src/webportal/src/app/vc/vc.component.scss]: interesting, I will add it back

File: src/rest-server/src/routes/authn.js (hunk 12 above)
Comment (ydye) [reply src/rest-server/src/routes/authn.js]: Sorry I don't think so.

File: src/rest-server/src/config/authn.js (hunk 3 above)
Comment (ydye) [reply src/rest-server/src/config/authn.js]: Done

File: src/rest-server/src/routes/v2/user.js (hunk 10 above)
Comment (ydye) [reply src/rest-server/src/routes/v2/user.js]: Done

File: src/rest-server/src/controllers/v2/azureAD.js (hunk 5 above)
Comment (ydye) [reply src/rest-server/src/controllers/v2/azureAD.js]: Done

**Pull request merged**